crates/dbsp/src/trace/spine_async/merger.rs

```diff
@@ -0,0 +1,150 @@
+use metrics::{counter, histogram};
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::Instant;
+
+pub(super) enum MergeOperation<B>
+where
+    B: Batch,
+{
+    Merge(
+        BatchIdent,
+        Vec<Arc<B>>,
+        Option<Filter<B::Key>>,
+        Option<Filter<B::Val>>,
+    ),
+}
+
+pub(super) enum MergeResult<B>
+where
+    B: Batch,
+{
+    MergeCompleted(Result<(BatchIdent, B), Error>),
+}
+
+pub(crate) struct BatchMerger<B>
+where
+    B: Batch,
+{
+    /// A handle to initiate merge operations on the merger thread.
+    sender: Sender<MergeOperation<B>>,
+    /// A handle to receive completed merge operations from the merger thread.
+    receiver: Receiver<MergeResult<B>>,
+    /// In progress merges.
+    _in_progress: Vec<(MergeOperation<B>, <B as Batch>::Merger)>,
+    /// To make sure we use the same batch type as the spine using this.
+    _phantom: PhantomData<B>,
+}
+
+impl<B> BatchMerger<B>
+where
+    B: Batch,
+{
+    /// We sample the queue length only once every `QUEUE_LENGTH_SAMPLE_RATE` merges.
+    const QUEUE_LENGTH_SAMPLE_RATE: usize = 1_000;
+
+    /// Size of the incoming merge queue.
+    const RX_QUEUE_SIZE: usize = 64;
+    /// How many concurrent merges we allow.
+    const CONCURRENT_MERGES: usize = 1;
+
+    pub(crate) fn new() -> Self {
+        let (new_merge_sender, new_merge_receiver) = bounded(Self::RX_QUEUE_SIZE);
+        let (completed_merge_sender, completed_merge_receiver) = unbounded();
+        let mut merge_cnt = 0;
+
+        Runtime::spawn_compactor(move || loop {
+            let op: Result<MergeOperation<B>, RecvError> = new_merge_receiver.recv();
+            merge_cnt += 1;
+            if merge_cnt % Self::QUEUE_LENGTH_SAMPLE_RATE == 0 {
+                histogram!(COMPACTION_DURATION).record(new_merge_receiver.len() as f64);
+            }
+
+            match op {
+                Ok(MergeOperation::Merge(key, mut batches, key_filter, value_filter)) => {
+                    // TODO: currently batches.len() is always 2, except when we call consolidate.
+                    // In the future we should make add support for multi-way merges so we don't
+                    // have to merge so inefficiently in presence of many batches.
+                    assert!(batches.len() > 1);
+
+                    let start = Instant::now();
+                    let mut fuel = isize::MAX;
+
+                    let mut current = batches.pop().unwrap();
+                    while let Some(next) = batches.pop() {
+                        let mut merger = <B as Batch>::Merger::new_merger(&current, &next);
+                        merger.work(&current, &next, &key_filter, &value_filter, &mut fuel);
+                        current = Arc::new(merger.done());
+                        counter!(TOTAL_COMPACTIONS).increment(1);
+                    }
+
+                    //assert!(current.len() > 0);
+                    histogram!(COMPACTION_SIZE).record(current.len() as f64);
```

Comment:
It might be useful also having a histogram or counter for how much the merge shortened the data (because of filters or because weights added to zero).